asyncpg/protocol/protocol.pyx

```diff
@@ -588,6 +588,8 @@ cdef class BaseProtocol(CoreProtocol):
             })
             self.abort()
 
+        if self.state == PROTOCOL_PREPARE:
+            self._write(SYNC_MESSAGE)
```

Comment:
It's probably safe to send `Sync` unconditionally here.

Reply:
It's not - some of the testcases fail when done unconditionally (I first had it like that). We get into a situation where we may end up in IDLE state due to the first sync received, and then receive another ReadyForQuery message - the current state machine goes into error state when that happens

Reply:
that may actually have to do with existing code line 120 - on receiving a Z message it doesn't parse the state. Isn't there supposed to be a:
`self._parse_msg_ready_for_query()`
on line 120?

Reply:
Yes, that is suspicious.  Server messages should always be read in full.  I'd say it's OK to also receive any number of `ReadyForQuery` in the `IDLE` state.  This should make it safe to send any number of `Sync` messages.

Reply:
I've experimented a bit with removing the if-statement and putting the ReadyForQuery handling in the state machine in IDLE state. However, it's giving me some warnings I'm not sure are good.
All test cases pass, but I see some extra lines like this after some test cases and I can't figure out exactly why it's not waiting for these.

```Future exception was never retrieved
future: <Future finished exception=BrokenPipeError(32, 'Broken pipe')>
Traceback (most recent call last):
  File "/usr/lib64/python3.6/asyncio/selector_events.py", line 417, in _sock_sendall
    n = sock.send(data)
BrokenPipeError: [Errno 32] Broken pipe```

I propose to keep this line as it is for now.